translate/src/scheduler.rs

```diff
@@ -1,77 +1,74 @@
 //! # harvest_translate scheduler
 //!
 //! The scheduler is responsible for determining which tools to invoke and also
-//! for invoking them. It is effectively the main loop of harvest_translate.
+//! for invoking them.
 
-use crate::ToolInvocation;
+use crate::Tool;
+use log::debug;
+use std::mem::replace;
 
 #[derive(Default)]
 pub struct Scheduler {
-    queued_invocations: Vec<ToolInvocation>,
+    queued_invocations: Vec<Box<dyn Tool>>,
 }
 
 impl Scheduler {
     /// Invokes `f` with the next suggested tool invocations. `f` is expected to try to run each
-    /// tool.
-    pub fn next_invocations<F: FnMut(&ToolInvocation) -> InvocationOutcome>(&mut self, mut f: F) {
-        self.queued_invocations.retain(|i| {
-            log::info!("Trying tool invocation {}", i);
-            let outcome = f(i);
-            log::info!("Tool invocation outcome: {outcome:?}");
-            outcome == InvocationOutcome::Wait
-        });
+    /// tool. If the tool cannot be executed and should be tried again later, then `f` should
+    /// return it.
+    pub fn next_invocations<F: FnMut(Box<dyn Tool>) -> Option<Box<dyn Tool>>>(&mut self, mut f: F) {
+        let new_queue = Vec::with_capacity(self.queued_invocations.len());
+        for tool in replace(&mut self.queued_invocations, new_queue) {
+            debug!("Trying to invoke tool {}", tool.name());
+            if let Some(tool) = f(tool) {
+                debug!("Returning {} to queue", tool.name());
+                self.queued_invocations.push(tool);
+            } else {
+                debug!("Tool removed from queue");
+            }
+        }
     }
 
     /// Add a tool invocation to the scheduler's queue. Note that scheduling a
     /// tool invocation does not guarantee the tool will run, as a tool may
     /// indicate that it is not runnable.
-    pub fn queue_invocation(&mut self, invocation: ToolInvocation) {
+    pub fn queue_invocation(&mut self, invocation: Box<dyn Tool>) {
```

Comment:
Is there a reason not to do the boxing _inside_ this function, so that it's signature is, say:

`pub fn queue_invocation<I: Tool>(&mut self, invocation: I)`

(or something)

so that the caller (`main.rs`) doesn't have to write `Box::new` each time?